client/relay-chain-interface/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use cumulus_primitives_core::{
+	relay_chain::{
+		v1::{CommittedCandidateReceipt, OccupiedCoreAssumption, SessionIndex, ValidatorId},
+		Block as PBlock, BlockId, Hash as PHash, InboundHrmpMessage,
+	},
+	InboundDownwardMessage, ParaId, PersistedValidationData,
+};
+use polkadot_overseer::Handle as OverseerHandle;
+use sc_client_api::{blockchain::BlockStatus, StorageProof};
+
+use sp_api::ApiError;
+use sp_core::sp_std::collections::btree_map::BTreeMap;
+use sp_state_machine::StorageValue;
+
+pub enum BlockCheckResult {
+	InChain,
+	NotFound(sc_client_api::ImportNotifications<PBlock>),
+}
+
+/// Should be used for all interaction with the relay chain in cumulus.
+pub trait RelayChainInterface: Send + Sync {
+	fn get_storage_by_key(&self, block_id: &BlockId, key: &[u8]) -> Option<StorageValue>;
+
+	fn validators(&self, block_id: &BlockId) -> Result<Vec<ValidatorId>, ApiError>;
+
+	fn block_status(&self, block_id: BlockId) -> Result<BlockStatus, sp_blockchain::Error>;
+
+	fn best_block_hash(&self) -> PHash;
+
+	/// Returns the whole contents of the downward message queue for the parachain we are collating
+	/// for.
+	///
+	/// Returns `None` in case of an error.
+	fn retrieve_dmq_contents(
+		&self,
+		para_id: ParaId,
+		relay_parent: PHash,
+	) -> Option<Vec<InboundDownwardMessage>>;
+
+	/// Returns channels contents for each inbound HRMP channel addressed to the parachain we are
+	/// collating for.
+	///
+	/// Empty channels are also included.
+	fn retrieve_all_inbound_hrmp_channel_contents(
+		&self,
+		para_id: ParaId,
+		relay_parent: PHash,
+	) -> Option<BTreeMap<ParaId, Vec<InboundHrmpMessage>>>;
+
+	fn persisted_validation_data(
+		&self,
+		block_id: &BlockId,
+		para_id: ParaId,
+		_: OccupiedCoreAssumption,
+	) -> Result<Option<PersistedValidationData>, ApiError>;
+
+	fn candidate_pending_availability(
+		&self,
+		block_id: &BlockId,
+		para_id: ParaId,
+	) -> Result<Option<CommittedCandidateReceipt>, ApiError>;
+
+	fn session_index_for_child(&self, block_id: &BlockId) -> Result<SessionIndex, ApiError>;
+
+	fn import_notification_stream(&self) -> sc_client_api::ImportNotifications<PBlock>;
+
+	/// Check if block is in chain. If it is, we return BlockCheckResult::InChain.
+	/// If it is not in the chain, we return BlockCheckResult::NotFound with a listener that can be used to wait on the block.
+	fn check_block_in_chain(
```

Comment:
Part of the docs should go to the `BlockCheckResult` type.

However, I'm not really happy with the naming here.

Maybe something like `wait_for_block` and return a future that either resolves directly or waits for the block to be imported.